packages/zoltan2/src/algorithms/connectivity/Zoltan2_IceSheet.hpp

```diff
@@ -0,0 +1,306 @@
+        int sendbufidx = sdispls[proc_to_send] + sentcount[proc_to_send];
+        sentcount[proc_to_send] += 2;
+        sendbuf[sendbufidx++] = boundary[i];
+        sendbuf[sendbufidx++] = boundary[i+1];
+      }
+    }
+ 
+    //Do the final communication back to each remote process 
+    Teuchos::ArrayRCP<gno_t> recvbuf;
+    Teuchos::ArrayView<const gno_t> sendBuf = Teuchos::arrayViewFromVector(sendbuf);
+    Zoltan2::AlltoAllv(*problemComm,*env,sendBuf,sendCount,recvbuf,recvCount);
+
+    //hash for the boundary-edge set
+    struct pair_hash {
+      inline std::size_t operator()(const std::pair<gno_t,gno_t>& v) const {
+        return v.first * 10 + v.second;
+      }
+    };
+
+    //create the set to see if we've counted this boundary edge before
+    std::unordered_set<std::pair<gno_t,gno_t>,pair_hash> edge_set;
+
+    int* local_boundary_counts = new int[nVtx];
+    for(size_t i = 0; i < nVtx; i++){
+      local_boundary_counts[i] = 0;
+    }
+  
+    //insert the local boundary edges into the set, small global endpoint first.
+    for(int i = 0; i < num_boundary_edges*2; i+=2){
+      if(owners[i] == me || owners[i+1] == me){
+        if(boundary_edges[i] < boundary_edges[i+1]){
+          edge_set.insert(std::make_pair(boundary_edges[i],boundary_edges[i+1]));
+        } else {
+          edge_set.insert(std::make_pair(boundary_edges[i+1],boundary_edges[i]));
+        }
+      }
+    }
+    //factor in the locally-owned boundary edges
+    for(int i = 0; i < num_boundary_edges*2; i++){
+      if(map->getLocalElement(boundary_edges[i]) != fail){
+        local_boundary_counts[map->getLocalElement(boundary_edges[i])]++;
+      }
+    }
+    
+    //use the received boundary edges, but check to make sure they haven't been used yet.
+    for(int i = 0; i < recvsize; i+=2){
+      //make sure the edge has not been used (if count returns 1 for one of these, the edge has been used)
+      if(edge_set.count(std::make_pair(recvbuf[i], recvbuf[i+1])) == 0 && edge_set.count(std::make_pair(recvbuf[i+1],recvbuf[i])) == 0){
+        //check which endpoints are owned locally, and set the boundary counts appropriately
+        if(map->getLocalElement(recvbuf[i]) != fail){
+          local_boundary_counts[map->getLocalElement(recvbuf[i])]++;
+        }
+        if(map->getLocalElement(recvbuf[i+1]) != fail){
+          local_boundary_counts[map->getLocalElement(recvbuf[i+1])]++;
+        }
+        //add the received edge to the set, to ensure no received duplicates are counted
+        if(recvbuf[i] < recvbuf[i+1]){
+          edge_set.insert(std::make_pair(recvbuf[i],recvbuf[i+1]));
+        } else {
+  	  edge_set.insert(std::make_pair(recvbuf[i+1],recvbuf[i]));
+        }
+      }
+    }
+    //we should now have complete local knowledge of the boundary.
+
+    //convert adjacency array to use local identifiers instead of global.
+  
+    typename map_t::local_ordinal_type* out_edges_lid = new typename map_t::local_ordinal_type[nEdge];
```

Comment:
The rest of the code uses std::vector or Teuchos::Array; why does this array use C++ array?

Reply:
It was due to the graph class having a C++ array as a member. I've fixed that, so the graph data structure uses Teuchos::Array, so this line has also been changed to Teuchos::Array.